dfinity/agent-rs
# feat: add fetch_api_boundary_nodes API to ic-agent

# Description

API boundary nodes were recently added to the certified state tree. Currently, there are no API boundary nodes and the subtree is empty. accordingly. API boundary nodes will be added to the registry via NNS proposals at some later stage.

IC clients should be able to retrieve all these API BNs in a convenient way.

This PR extends `ic-agent` API with a new method:
`pub async fn fetch_api_boundary_nodes() -> Vec<ApiBoundaryNode>`

# How Has This Been Tested?

1. Run this [api_boundary_node_decentralization_test](https://sourcegraph.com/github.com/dfinity/ic@db9595a8be755949a28e0af74c6bae85fa1a9c3d/-/blob/rs/tests/boundary_nodes/BUILD.bazel?L142) `system-test`. This test converts two unassigned nodes into API BNs.
2. Instantiate an `agent` with any IC replica url provided in `system-test` logs during test execution.
3. Retrieve API BNs via `let api_bns = agent.fetch_api_boundary_nodes(effective_canister_id).await`.
4. Verify that two  API BNs are discovered within the `State Tree`.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
